---
title: "Analyzing MeRIP-seq data with MeRIPeak"
author: "Zhenxing Guo"
date: "1/7/2020"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Put the title of your vignette here}
  %\VignetteEngine{knitr::rmarkdown}
  \usepackage[utf8]{inputenc}
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Abstract
The post-transcriptional epigenetic modiﬁcation on mRNA is an emerging ﬁeld to study the gene regulatory mechanism and their association with diseases. Recently developed high-throughput sequencing technology named Methylated RNA Immunoprecipitation Sequencing (MeRIP-seq) enables one to proﬁle mRNA epigenetic modiﬁcation transcriptome-wide. A basic task in the analysis of MeRIP-seq data is to identify transcriptome-wide m6A regions (namely "peak calling"). The package MeRIPeak provides methods for peak calling of MeRIP-seq data, based on an empirical Bayesian hierarchical model. The method accounts for various sources of variations in the data through rigorous modeling, and achieves shrinkage estimation by borrowing informations from transcriptome-wide data to stabilize the parameter estimation. This vignette explains the use of the package by introducing typical workflows. MeRIPeak package version: 0.1.0.

## Acknowledgments

## Running MeRIPeak
### 1. Installation
From GitHub: 
```{r, eval = FALSE}
install.packages("devtools") # if you have not installed "devtools" package
library(devtools)
install_github("https://github.com/ZhenxingGuo0015/MeRIPeak")
```

### A quick example to call peaks
It is very convenient to conduct peak calling with the following code chunk. All you need are BAM files of IP and input samples output from sequence alignment tools like Bowtie2.
The following code assumes you know (of course) the reference genome  "Refgenome" ( like "mm9", "mm10", "hg18", "hg19",...) of your sequencing data. "BamDir" is the directory where you store the BAM files for both IP ("ip1.bam", "ip2.bam", ...) and input ("input1.bam", "input2.bam", ...) samples. "OutDir" and "NameofPeak" are your output directory and the name of your assigned for the excel file containing called peaks respectively.

```{r, eval = FALSE}
IP.file = c("ip1.bam", "ip2.bam")
Input.file = c("input1.bam", "input2.bam")
MeRiPeak(IP.file = IP.file,
         Input.file = Input.file,
         genomeBuild = Refgenome,
         InputDir = BamDir,
         OutputDir = OutDir,
         experiment_name = NameofResults)
```

There are one main function and multiple subfunctions in MeRIPeak for different usages.

## 2. Main function: MeRiPeak
### Parameters



## 3. Usage of different sub functions.
### 3.1 Call peaks starting from bin-level count matrix input 
If you have more than 2 replicates and already have bin-level counts from BAM files. You can use functions "M6Apeak.MultiRep.step1" and "M6Apeak.MultiRep.step2" for peak calling.

#### M6Apeak.MultiRep.step1
##### Parameters
#### M6Apeak.MultiRep.step2
##### Parameters

### 3.2 Call peaks starting from region-level count matrix input
If you already have a list of peaks and the read counts but you want to re-rank them using MeRIPeak. You can use function "M6Apeak" to do this.

#### M6Apeak
##### Parameters

### 3.3 Call peaks with only one replicate for both IP and input samples.
If there is only one biological replicate from your experiment. You can use function "M6Apeak.oneRep" to conduct peak calling.

#### M6Apeak.oneRep
##### Parameters
