<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />

<meta name="viewport" content="width=device-width, initial-scale=1">

<meta name="author" content="Zhenxing Guo" />


<title>Analyzing MeRIP-seq data with TRES</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' || rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>



<link href="data:text/css;charset=utf-8,body%20%7B%0Abackground%2Dcolor%3A%20%23fff%3B%0Amargin%3A%201em%20auto%3B%0Amax%2Dwidth%3A%20700px%3B%0Aoverflow%3A%20visible%3B%0Apadding%2Dleft%3A%202em%3B%0Apadding%2Dright%3A%202em%3B%0Afont%2Dfamily%3A%20%22Open%20Sans%22%2C%20%22Helvetica%20Neue%22%2C%20Helvetica%2C%20Arial%2C%20sans%2Dserif%3B%0Afont%2Dsize%3A%2014px%3B%0Aline%2Dheight%3A%201%2E35%3B%0A%7D%0A%23header%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0A%23TOC%20%7B%0Aclear%3A%20both%3B%0Amargin%3A%200%200%2010px%2010px%3B%0Apadding%3A%204px%3B%0Awidth%3A%20400px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Aborder%2Dradius%3A%205px%3B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Afont%2Dsize%3A%2013px%3B%0Aline%2Dheight%3A%201%2E3%3B%0A%7D%0A%23TOC%20%2Etoctitle%20%7B%0Afont%2Dweight%3A%20bold%3B%0Afont%2Dsize%3A%2015px%3B%0Amargin%2Dleft%3A%205px%3B%0A%7D%0A%23TOC%20ul%20%7B%0Apadding%2Dleft%3A%2040px%3B%0Amargin%2Dleft%3A%20%2D1%2E5em%3B%0Amargin%2Dtop%3A%205px%3B%0Amargin%2Dbottom%3A%205px%3B%0A%7D%0A%23TOC%20ul%20ul%20%7B%0Amargin%2Dleft%3A%20%2D2em%3B%0A%7D%0A%23TOC%20li%20%7B%0Aline%2Dheight%3A%2016px%3B%0A%7D%0Atable%20%7B%0Amargin%3A%201em%20auto%3B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dcolor%3A%20%23DDDDDD%3B%0Aborder%2Dstyle%3A%20outset%3B%0Aborder%2Dcollapse%3A%20collapse%3B%0A%7D%0Atable%20th%20%7B%0Aborder%2Dwidth%3A%202px%3B%0Apadding%3A%205px%3B%0Aborder%2Dstyle%3A%20inset%3B%0A%7D%0Atable%20td%20%7B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dstyle%3A%20inset%3B%0Aline%2Dheight%3A%2018px%3B%0Apadding%3A%205px%205px%3B%0A%7D%0Atable%2C%20table%20th%2C%20table%20td%20%7B%0Aborder%2Dleft%2Dstyle%3A%20none%3B%0Aborder%2Dright%2Dstyle%3A%20none%3B%0A%7D%0Atable%20thead%2C%20table%20tr%2Eeven%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Ap%20%7B%0Amargin%3A%200%2E5em%200%3B%0A%7D%0Ablockquote%20%7B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Apadding%3A%200%2E25em%200%2E75em%3B%0A%7D%0Ahr%20%7B%0Aborder%2Dstyle%3A%20solid%3B%0Aborder%3A%20none%3B%0Aborder%2Dtop%3A%201px%20solid%20%23777%3B%0Amargin%3A%2028px%200%3B%0A%7D%0Adl%20%7B%0Amargin%2Dleft%3A%200%3B%0A%7D%0Adl%20dd%20%7B%0Amargin%2Dbottom%3A%2013px%3B%0Amargin%2Dleft%3A%2013px%3B%0A%7D%0Adl%20dt%20%7B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0Aul%20%7B%0Amargin%2Dtop%3A%200%3B%0A%7D%0Aul%20li%20%7B%0Alist%2Dstyle%3A%20circle%20outside%3B%0A%7D%0Aul%20ul%20%7B%0Amargin%2Dbottom%3A%200%3B%0A%7D%0Apre%2C%20code%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0Aborder%2Dradius%3A%203px%3B%0Acolor%3A%20%23333%3B%0Awhite%2Dspace%3A%20pre%2Dwrap%3B%20%0A%7D%0Apre%20%7B%0Aborder%2Dradius%3A%203px%3B%0Amargin%3A%205px%200px%2010px%200px%3B%0Apadding%3A%2010px%3B%0A%7D%0Apre%3Anot%28%5Bclass%5D%29%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Acode%20%7B%0Afont%2Dfamily%3A%20Consolas%2C%20Monaco%2C%20%27Courier%20New%27%2C%20monospace%3B%0Afont%2Dsize%3A%2085%25%3B%0A%7D%0Ap%20%3E%20code%2C%20li%20%3E%20code%20%7B%0Apadding%3A%202px%200px%3B%0A%7D%0Adiv%2Efigure%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0Aimg%20%7B%0Abackground%2Dcolor%3A%20%23FFFFFF%3B%0Apadding%3A%202px%3B%0Aborder%3A%201px%20solid%20%23DDDDDD%3B%0Aborder%2Dradius%3A%203px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Amargin%3A%200%205px%3B%0A%7D%0Ah1%20%7B%0Amargin%2Dtop%3A%200%3B%0Afont%2Dsize%3A%2035px%3B%0Aline%2Dheight%3A%2040px%3B%0A%7D%0Ah2%20%7B%0Aborder%2Dbottom%3A%204px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Apadding%2Dbottom%3A%202px%3B%0Afont%2Dsize%3A%20145%25%3B%0A%7D%0Ah3%20%7B%0Aborder%2Dbottom%3A%202px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Afont%2Dsize%3A%20120%25%3B%0A%7D%0Ah4%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23f7f7f7%3B%0Amargin%2Dleft%3A%208px%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Ah5%2C%20h6%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23ccc%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Aa%20%7B%0Acolor%3A%20%230033dd%3B%0Atext%2Ddecoration%3A%20none%3B%0A%7D%0Aa%3Ahover%20%7B%0Acolor%3A%20%236666ff%3B%20%7D%0Aa%3Avisited%20%7B%0Acolor%3A%20%23800080%3B%20%7D%0Aa%3Avisited%3Ahover%20%7B%0Acolor%3A%20%23BB00BB%3B%20%7D%0Aa%5Bhref%5E%3D%22http%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0Aa%5Bhref%5E%3D%22https%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0A%0Acode%20%3E%20span%2Ekw%20%7B%20color%3A%20%23555%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Edt%20%7B%20color%3A%20%23902000%3B%20%7D%20%0Acode%20%3E%20span%2Edv%20%7B%20color%3A%20%2340a070%3B%20%7D%20%0Acode%20%3E%20span%2Ebn%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Efl%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Ech%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Est%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Eco%20%7B%20color%3A%20%23888888%3B%20font%2Dstyle%3A%20italic%3B%20%7D%20%0Acode%20%3E%20span%2Eot%20%7B%20color%3A%20%23007020%3B%20%7D%20%0Acode%20%3E%20span%2Eal%20%7B%20color%3A%20%23ff0000%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Efu%20%7B%20color%3A%20%23900%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%20code%20%3E%20span%2Eer%20%7B%20color%3A%20%23a61717%3B%20background%2Dcolor%3A%20%23e3d2d2%3B%20%7D%20%0A" rel="stylesheet" type="text/css" />




</head>

<body>




<h1 class="title toc-ignore">Analyzing MeRIP-seq data with TRES</h1>
<h4 class="author">Zhenxing Guo</h4>



<div id="abstract" class="section level2">
<h2>Abstract</h2>
<p>The post-transcriptional epigenetic modiﬁcation on mRNA is an emerging ﬁeld to study the gene regulatory mechanism and their association with diseases. Recently developed high-throughput sequencing technology named Methylated RNA Immunoprecipitation Sequencing (MeRIP-seq) enables one to proﬁle mRNA epigenetic modiﬁcation transcriptome-wide. A basic task in the analysis of MeRIP-seq data is to identify transcriptome-wide m6A regions (namely “peak calling”). The package TRES provides methods for peak calling of MeRIP-seq data, based on an empirical Bayesian hierarchical model. The method accounts for various sources of variations in the data through rigorous modeling, and achieves shrinkage estimation by borrowing informations from transcriptome-wide data to stabilize the parameter estimation. This vignette explains the use of the package by introducing typical workflows. TRES package version: 0.1.0.</p>
</div>
<div id="acknowledgments" class="section level2">
<h2>Acknowledgments</h2>
</div>
<div id="running-tres-for-peak-calling" class="section level2">
<h2>Running TRES for peak calling</h2>
<div id="installation" class="section level3">
<h3>1. Installation</h3>
<p>From GitHub:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">install.packages</span>(<span class="st">&quot;devtools&quot;</span>) <span class="co"># if you have not installed &quot;devtools&quot; package</span>
<span class="kw">library</span>(devtools)
<span class="kw">install_github</span>(<span class="st">&quot;https://github.com/ZhenxingGuo0015/TRES&quot;</span>)</code></pre></div>
<p>There are one main function and multiple subfunctions in TRES for different usages.</p>
</div>
</div>
<div id="main-function-of-peak-calling-tres_peak" class="section level2">
<h2>2. Main function of peak calling: TRES_peak</h2>
<p>It is very convenient to conduct peak calling using this function. It starts from BAM files, so all you need are BAM files of IP and input samples output from sequence alignment tools like Bowtie2. It requires you know (of course) the reference genome ( like “mm9”, “mm10”, “hg18”, “hg19”,…) of your sequencing data. “InputDir” is the directory where you store the BAM files for both IP (“ip1.bam”, “ip2.bam”, …) and input (“input1.bam”, “input2.bam”, …) samples. “OutDir” and “experiment_name” are your output directory and the name you want for the peak excel file respectively.</p>
<p>A quick example with setup of required parameters:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">IP.file =<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;ip1.bam&quot;</span>, <span class="st">&quot;ip2.bam&quot;</span>)
Input.file =<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;input1.bam&quot;</span>, <span class="st">&quot;input2.bam&quot;</span>)
MyInputDir =<span class="st"> &quot;/Users/zhenxingguo/Documents/research/m6a/Me/working/RealdataAnalysis/data/Mouse/bam/&quot;</span>
MyOutDir =<span class="st"> &quot;/Users/zhenxingguo/Documents/research/m6a/Me/working/RealdataAnalysis/Results/Mouse/PeakList/MyPeak/&quot;</span>
<span class="kw">TRES_peak</span>(<span class="dt">IP.file =</span> IP.file,
         <span class="dt">Input.file =</span> Input.file,
         <span class="dt">genomeBuild =</span> <span class="st">&quot;mm9&quot;</span>,
         <span class="dt">InputDir =</span> MyInputDir,
         <span class="dt">OutputDir =</span> MyOutDir,
         <span class="dt">experiment_name =</span> <span class="st">&quot;example&quot;</span>)

peaks =<span class="st"> </span><span class="kw">read.table</span>(<span class="kw">paste0</span>(MyOutDir, <span class="st">&quot;example_peaks.xls&quot;</span>), <span class="dt">sep =</span> <span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>, <span class="dt">header =</span> <span class="ot">TRUE</span>)
<span class="kw">head</span>(peaks)</code></pre></div>
<div id="tres_peak-parameters" class="section level3">
<h3>TRES_peak Parameters</h3>
<p><strong>IP.file</strong>: A vector of characters containing the name of bam files for IP samples.</p>
<p><strong>Input.file</strong>: A vector of characters containing the name of bam files for input control samples.</p>
<p><strong>genomeBuild</strong>: A character to specify the reference genome of the sequencing data, which takes value in (“mm9”, “mm10”, “hg18”, “hg19”, “dm3”, “rn4”).</p>
<p><strong>binsize</strong>: A numerical value to specify the size of window to bin the genome and get bin-level read counts. Default value is 50.</p>
<p><strong>sf0</strong>: Numerical vectors to specify the size factors of each sample. Default is NULL.</p>
<p><strong>WhichThreshold</strong>: A character to specify the name of threshod to screen candidate regions in the first step. It takes among “pval”, “fdr”, “lfc”, “pval_lfc” and “fdr_lfc”. “pval”: screen canidates only based on P-values from Binomial tests; “fdr”: screen canidates only based on FDR; “lfc”: screen canidates only based on log fold changes between normalized IP and normalized input read counts; “pval_lfc”: screen canidates based on both P-values and log fold changes; “fdr_lfc”: screen canidates based on both FDR and log fold changes. Default value of ``WhichThreshold&quot; is “fdr”.</p>
<p><strong>pval.cutoff0</strong>: A numerical value to specify the cutoff of p-values if identifying the candidate regions based on P-values. Default is 1e-5.</p>
<p><strong>fdr.cutoff0</strong>: A numerical value to specify the cutoff of fdr if identifying the candidate regions based on FDR. Default is 1e-5.</p>
<p><strong>lfc.cutoff0</strong>: A numerical value to specify the cutoff of log foldchanges between normalized IP and input read counts if identifying the candidate regions based on log fold changes. Default is 0.7.</p>
<p><strong>mu.cutoff0</strong>: A numerical value to specify the background methylation level. Default is 0.5.</p>
<p><strong>addDEseq2</strong>: A logical value to indicate whether or not include DESeq2 in the second step of ranking candidate regions. Default is FALSE.</p>
<p><strong>InputDir</strong>: A character to specify the input directory of all bam files.</p>
<p><strong>OutputDir</strong>: A character to specify the output directory of all results.</p>
<p><strong>experiment_name</strong>: A character to specify the name of results folder.</p>
</div>
<div id="tres_peak-outputs" class="section level3">
<h3>TRES_peak Outputs</h3>
<p>The output of this function contains two set of results. One is saved as “.rda”, which contains all the genomic bins from the reference genome and theirs read counts. The other is one “.xls” file, which contains a list of called peaks. The columns of the peak excel files are:</p>
<p><strong>chr</strong>: Chromosome number.</p>
<p><strong>start</strong>: The start of genomic position of that peak.</p>
<p><strong>end</strong>: The end of genomic position of that peak.</p>
<p><strong>strand</strong>: The strand of genomic position of that peak.</p>
<p><strong>summit</strong>: The summit position of that peak.</p>
<p><strong>lg.fc</strong>: The log foldchange between normalized IP and normalized input read counts.</p>
<p><strong>mu</strong>: The methylation level of that peak if there are more than one replicate.</p>
<p><strong>mu.var</strong>: The estimated variance of for methylation level of that peak, if there are more than one replicate.</p>
<p><strong>stat</strong>: The Wald test statistics of that peak, if there are more than one replicate.</p>
<p><strong>shrkPhi</strong>: The shrinkage estimation of dispersion for mehtylation levels of that peak, if there are more than one replicate.</p>
<p><strong>shrkTheta</strong>: The shrinkage estimation for scale parameter theta in the gamma distribution, if there are more than one replicate.</p>
<p><strong>pvals</strong>: P-value calculated based on the Wald-test.</p>
<p><strong>score</strong>: Adjusted p-values using FDR.</p>
<p>Note, there are additional columns whose name involves the character “.bam”. These columns contain the peak-level read counts in respective samples.</p>
</div>
</div>
<div id="usage-of-different-sub-functions-for-peak-calling." class="section level2">
<h2>3. Usage of different sub functions for peak calling.</h2>
<div id="call-peaks-starting-from-bin-level-count-matrix-input" class="section level3">
<h3>3.1 Call peaks starting from bin-level count matrix input</h3>
<p>If you have more than 1 replicates and already have bin-level counts from BAM files. You can use functions “M6Apeak.MultiRep.step1” and “M6Apeak.MultiRep.step2” togther for peak calling. “M6Apeak.MultiRep.step1” detect and combine significant bins to form candidate regions and “M6Apeak.MultiRep.step2” ranks candidate regions separately by the proposed empirical Bayesian hirarchical negative binomial model and by DESeq2 (if required).</p>
<p>A quick example:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">data</span>(<span class="st">&quot;Basal_binlevel&quot;</span>) 
### The first 14 columns are from basal samples of mouse cortex
sf0 =<span class="st"> </span><span class="kw">colSums</span>(allbincount)<span class="op">/</span><span class="kw">median</span>(<span class="kw">colSums</span>(allbincount))
peak_step1 =<span class="st"> </span><span class="kw">M6Apeak.MultiRep.step1</span>(<span class="dt">Counts =</span> allbincount[, <span class="dv">1</span><span class="op">:</span><span class="dv">14</span>],
                                    <span class="dt">bins =</span> allbins,
                                    <span class="dt">sf =</span> sf0[<span class="dv">1</span><span class="op">:</span><span class="dv">14</span>],
                                    <span class="dt">WhichThreshold =</span> <span class="st">&quot;lfc&quot;</span>,
                                    <span class="dt">lfc.cutoff =</span> <span class="fl">0.5</span>)
Peaks =<span class="st"> </span><span class="kw">M6Apeak.MultiRep.step2</span>(<span class="dt">Candidates =</span> peak_step1, 
                               <span class="dt">sf =</span> sf0[<span class="dv">1</span><span class="op">:</span><span class="dv">14</span>])
<span class="kw">head</span>(Peaks)</code></pre></div>
<div id="m6apeak.multirep.step1" class="section level4">
<h4>M6Apeak.MultiRep.step1</h4>
<p>This function assumes you have already binned the whole reference transcriptome into windows of fixed length (such as 50 basepair long) and obtained the corresponding read counts within each window for all samples.</p>
<div id="parameters" class="section level5">
<h5>Parameters</h5>
<p><strong>Counts</strong>: A data matrix containing bin-level (50bp) read counts in both IP and input samples. The number of column depends on the number of replicates, where the sample order is: input1, ip1, input2, ip2, …</p>
<p><strong>sf</strong>: A vector specifying the size factor of each sample, which is estimated using “Counts”: colSums(Counts)/median(colSums(Counts)), or provided by the user.</p>
<p><strong>bins</strong>: A data frame containing the genomic position of each bin of length 50bp.</p>
<p><strong>WhichThreshold</strong>: A character specifying a threshold for significant bins in bump finding using an ad hoc algorithm. There are three options: “pval” (only use p-values), “fdr” (only use FDR), “lfc” (only use log fold change), “pval_lfc” (use both p-values and log fold changes) and “fdr_lfc” (use FDR and log fold changes). Default is “fdr”.</p>
<p><strong>pval.cutoff</strong>: A constant indicating the cutoff for p-value. Default is 1e-05.</p>
<p><strong>fdr.cutoff</strong>: A constant indicating the cutoff for FDR. Default is 1e-05.</p>
<p><strong>lfc.cutoff</strong>: A constant indicating the cutoff for log fold change. Default is 0.7 for fold change of 2.</p>
<p><strong>windlen</strong>: An integer specifying the length of consecutive bins used in simple moving average smooth of log fold change.</p>
<p><strong>lowcount</strong>: An iteger to filter out candidate regions with lower read counts. Default is 10.</p>
</div>
<div id="outputs" class="section level5">
<h5>Outputs</h5>
<p>The output of this function is a dataframe containing candidate peak information denoted by the following columns:</p>
<p><strong>chr</strong>: Chromosome number.</p>
<p><strong>start</strong>: Start position.</p>
<p><strong>end</strong>: End position.</p>
<p><strong>summit</strong>: The summit of each peak.</p>
<p><strong>score</strong>: The adjusted P-value of each candidate region. The P-value is derived using on the binomial test based on the region-level read counts.</p>
<p><strong>lg.fc</strong>: The log fold changes between IP and input read counts within each m6A region.</p>
<p><strong>Others</strong>: All the other columns corresponds to read counts of IP/input within each region from each sample.</p>
</div>
</div>
<div id="m6apeak.multirep.step2" class="section level4">
<h4>M6Apeak.MultiRep.step2</h4>
<div id="parameters-1" class="section level5">
<h5>Parameters</h5>
<p><strong>Candidates</strong>: A dataframe which contain genomic location and read counts of candidate m6A regions.</p>
<p><strong>sf</strong>: A vector of size factors for each sample, which is the same as that used in the first step of identifying candidate regions.</p>
<p><strong>mu.cutoff</strong>: A constant specifying the background methylation levels. Default is 0.5.</p>
<p><strong>addDEseq2</strong>: A logical value, which indicate whether also ranking candidate regions using DESeq2 by treat the IP and input as paired data. Default is FALSE.</p>
</div>
<div id="outputs-1" class="section level5">
<h5>Outputs</h5>
<p>The output is a dataframe. In addition to the genomic locations, read counts and log foldchanges that are output by “M6Apeak.MultiRep.step1”, it also contains:</p>
<p><strong>mu</strong>: The methylation level of that peak if there are more than one replicate.</p>
<p><strong>mu.var</strong>: The estimated variance of for methylation level of that peak, if there are more than one replicate.</p>
<p><strong>stats</strong>: The Wald test statistics of that peak, if there are more than one replicate.</p>
<p><strong>shrkPhi</strong>: The shrinkage estimation of dispersion for mehtylation levels of that peak, if there are more than one replicate.</p>
<p><strong>shrkTheta</strong>: The shrinkage estimation for scale parameter theta in the gamma distribution, if there are more than one replicate.</p>
<p><strong>pvals</strong>: P-value calculated based on the Wald-test.</p>
<p><strong>score</strong>: Adjusted p-values using FDR.</p>
<p>Note, if you also want the rank regions by DESeq2 and set “addDEseq2 == TRUE”, then there would be an extra prefix “NB_” in the name of above output arguments. Also, there would be another three columns from DESeq2: “DESeq_pval”, “DESeq_score” and “DESeq_stat”.</p>
<p>If there is only one biological replicate from your experiment. You can use function <strong>M6Apeak.oneRep</strong> to conduct peak calling, which also starts with bin-level read counts as input. The input and output of this function are pretty much similar to the input of “M6Apeak.MultiRep.step1”.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># A toy example</span>
<span class="kw">data</span>(<span class="st">&quot;Basal_binlevel&quot;</span>)
peaks =<span class="st"> </span><span class="kw">M6Apeak.oneRep</span>(<span class="dt">Counts =</span> allbincount[, <span class="dv">1</span><span class="op">:</span><span class="dv">2</span>], <span class="dt">sf =</span> sf0[<span class="dv">1</span><span class="op">:</span><span class="dv">2</span>], <span class="dt">bins =</span> allbins)
<span class="kw">head</span>(peaks)</code></pre></div>
</div>
</div>
</div>
<div id="other-usage" class="section level3">
<h3>3.2 Other usage</h3>
<p>If you already have a list of peaks and the read counts but you want to re-rank them using TRES. You can also use function “M6Apeak.MultiRep.step2” to do this. This may perform bad if you didn’t appropriately estiamte size factors. Based on our experience, the estimation of size factor should be based on the bin-level counts across the transcriptome, instead of region-level counts.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">data</span>(<span class="st">&quot;Basal_regionlevel&quot;</span>)
<span class="kw">data</span>(<span class="st">&quot;Basal_binlevel&quot;</span>) ###  sf0 estimated from bin-level count
peaks =<span class="st"> </span><span class="kw">M6Apeak.MultiRep.step2</span>(<span class="dt">Candidates =</span> Basal, <span class="dt">sf =</span> sf0, <span class="dt">addDEseq2 =</span> <span class="ot">TRUE</span>)
<span class="kw">head</span>(peaks)</code></pre></div>
</div>
<div id="code-to-recover-part-of-results-in-the-manuscript" class="section level3">
<h3>4. Code to recover part of results in the manuscript</h3>
</div>
</div>



<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
